### ISSUE TITLE
Improve experience with screenreaders

### ISSUE THREAD

**Issue by Jermolene** [Jermolene/TiddlyWiki5#650]
There are several issues:
- Toolbar buttons have an icon but lack a label so the screenreader just says the word "button"
- The close buttons in the open tab of the sidebar lack a label so the screenreader just says the letter "x"

**Comment by tobibeer** [comment]
What screenreaders (or spec thereof) are we addressing?

**Comment by Jermolene** [comment]
I think that this ticket was prompted by feedback from Simon Eigeldinger via the mailing list. As I remember it, he uses two different screenreaders.

OS X has a built-in screenreader which I've also experimented with myself.

**Comment by domasofan** [comment]
hi,

i tested that on windows using the 2 screenreaders NVDA which is an open 
source one and JAWS which is the widely used very expensive one.
i also played with it under linux using the orca screen reader which is 
preinstalled on gnome.

greetings,
simon


Am 09.02.2015 um 19:27 schrieb Tobias Beer:
> What screenreaders (or spec thereof) are we addressing?
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/Jermolene/TiddlyWiki5/issues/650#issuecomment-73561644
>

-- 
Simon Eigeldinger
Follow me on Twitter: http://www.twitter.com/domasofan/
E-Mail: simon.eigeldinger@vol.at
MSN: simon_eigeldinger@hotmail.com
ICQ: 121823966
Jabber: domasofan@andrelouis.com

---
Diese E-Mail wurde von Avast Antivirus-Software auf Viren geprüft.
http://www.avast.com

**Comment by Flameborn** [comment]
I am currently testing e84c87ef3 built via the *build-site.sh* script under Safari, via VoiceOver. Most of the issues I found, however, will be generally present when using other screen readers as well, such as NVDA, Jaws, Narrator, Orca, etc.

* Controls that cause sections to appear, e.g. the *More*, or the *tag list* button need an [aria-expanded](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role) attribute, which is true when the section is shown, and becomes false when the section is hidden.
* It seems like the tags are automatically shown by default, or when the *Tag* field is focused. While this is great visually, a screen reader user needs to click the *tag list* button afterwards, since there is no way to skip past the list of tags quickly. When the *tag* field loses focus, the list of tags should be hidden, so that showing/hiding tags is only controlled by either focusing the field, which allows clicking on a tag visually, or via the *tag list* button, where the list of tags can be navigated to.
* The available tags do not seem to have a control type defined. Without looking at the source, I am guessing that they are a list of *div* or *span* containers. Screen readers only announce that these tags are clickable, but they can't be focused via the tab key, nor they are conveyed as a button (which essentially they are). Adding a *role* ARIA attribute defined as a button, as well as setting the *tabIndex* attribute to -1 should make these items focusable via the keyboard as well.
* I am not sure how customizable the default TiddlyWiki editor is, in case it is: The editor toolbar widgets should be contained in an ARIA [Toolbar](https://www.w3.org/TR/wai-aria-practices/#toolbar), so that screen readers can display these controls as part of a single group. Under MacOS, this allows skipping past all the toolbar items, as opposed to having to navigate through every single widget to reach the edit textbox.
* It's worth noting that the *More Actions* button of the editor needs an *aria-expanded* attribute as well.
* The *content type* and the *field* list behaves the same way as the list of tags, see above.
* Items in a group that can be selected to display their content, such as in the case of the control panel buttons, [aria-current](https://www.aditus.io/aria/aria-current/) should be used. In this case, you could add a *tab* role for the button, or even use *aria-selected*, so that a screen reader user knows immediately which page is active, rather than having to deduct it from the content.
* The button for expanding a keyboard shortcut (Control Panel -> Keyboard Shortcuts) does not have a text label. The same is also true for defining a new shortcut, which I assume is a new/edit button for the various platforms.

**Comment by Jermolene** [comment]
Thank you @Flameborn, and my apologies for the late response. I really appreciate the feedback. There's no reason for TiddlyWiki not to be accessible to screen readers, apart from my ignorance, so I'm very keen to learn more.

I think it might be easier to move your points to a new thread; this ancient one was prompted by a stream of feedback I received via email from a particular screenreader user. I've copied and pasted it to https://github.com/Jermolene/TiddlyWiki5/issues/5519
